### PR TITLE
Fix Teleport ending trainer battles

### LINKED ISSUE #3164: AI using Teleport in a trainer battle ends the battle
### Description

If the AI uses teleport in a trainer battle, the battle is ended like it were a wild battle.
teleport

### Version

upcoming (Edge)

### Upcoming Version

_No response_

### Discord contact info

joggel

### PR DESCRIPTION
**Pull request by AlexOn1ine**
## Description
Fix by redirecting to the Baton Pass script when it's a trainer battle.
## Issue(s) that this PR fixes
Fixes  #3164

### CODE REVIEW

**Comment by AsparagusEduardo** [comment]
Redirecting to Baton Pass' script will reduce PP twice

**Comment by LOuroboros** [comment]
You should update `CanTeleport` directly imo, @AlexOn1ine.

**Comment by AsparagusEduardo** [comment]
> ```diff
>     switch (GetBattlerSide(battlerId))
>     {
>     case B_SIDE_OPPONENT:
> -       if (count == 1 || gBattleTypeFlags & BATTLE_TYPE_DOUBLE)
> +       if (count == 1 || gBattleTypeFlags & (BATTLE_TYPE_DOUBLE | BATTLE_TYPE_TRAINER))
>             return FALSE;
>         break;
> ```

This implies that Teleport fails in trainer battles for the opponent, which is not the case.

**Comment by AlexOn1ine** [comment]
> Redirecting to Baton Pass' script will reduce PP twice

Fixed

> EDIT2: Also, I intended to do this at some point, but I guess I never ended up pushing it. canteleport and the subsequent jumpifbyte could be turned into a macro like jumpifcanteleport or something like that.

I think `canteleport` isn't needed anymore so I removed it completely.

**Comment by LOuroboros** [comment]
> > Redirecting to Baton Pass' script will reduce PP twice
> 
> Fixed
> 
> > EDIT2: Also, I intended to do this at some point, but I guess I never ended up pushing it. canteleport and the subsequent jumpifbyte could be turned into a macro like jumpifcanteleport or something like that.
> 
> I think `canteleport` isn't needed anymore so I removed it completely.

Noice!

The suggestion can still be applied to `getbattlerside` which could be used for a `jumpifside` macro to save a few lines in the file though :eyes:

**Comment by AlexOn1ine** [comment]
> The suggestion can still be applied to getbattlerside which could be used for a jumpifside macro to save a few lines in the file though eyes

Something like this? 
```diff
+	.macro jumpifside battler:req, side:req, jumpInstr:req
+	getbattlerside \battler
+	jumpifbyte CMP_EQUAL, gBattleCommunication, \side, \jumpInstr
+	.endm

-       getbattlerside BS_ATTACKER
-       jumpifbyte CMP_EQUAL, gBattleCommunication, B_SIDE_PLAYER, BattleScript_EffectBatonPass
+       jumpifside BS_ATTACKER, B_SIDE_PLAYER, BattleScript_EffectBatonPass
        jumpifbattletype BATTLE_TYPE_TRAINER, BattleScript_EffectBatonPass
-       jumpifbyte CMP_EQUAL, gBattleCommunication, B_SIDE_OPPONENT, BattleScript_EffectTeleportTryToRunAway
+       jumpifside BS_ATTACKER, B_SIDE_OPPONENT, BattleScript_EffectTeleportTryToRunAway
```

**Comment by LOuroboros** [comment]
> Something like this?

Yup. I guess that adds a pointless function call to `getbattlerside` though, which could be worked around like this:

```diff
+	.macro jumpifside battler:req, side:req, equalJumpInstr:req, notEqualJumpInstr:req
+	getbattlerside \battler
+	jumpifbyte CMP_EQUAL, gBattleCommunication, \side, \equalJumpInstr
+	jumpifbyte CMP_NOT_EQUAL, gBattleCommunication, \side, \notEqualJumpInstr
+	.endm

-       getbattlerside BS_ATTACKER
-       jumpifbyte CMP_EQUAL, gBattleCommunication, B_SIDE_PLAYER, BattleScript_EffectBatonPass
+       jumpifside BS_ATTACKER, B_SIDE_PLAYER, BattleScript_EffectBatonPass, BattleScript_EffectTeleportTryToRunAway
        jumpifbattletype BATTLE_TYPE_TRAINER, BattleScript_EffectBatonPass
-       jumpifbyte CMP_EQUAL, gBattleCommunication, B_SIDE_OPPONENT, BattleScript_EffectTeleportTryToRunAway
```

**Comment by AlexOn1ine** [comment]
> Yup. I guess that adds a pointless function call to getbattlerside though, which could be worked around like this:

Looks good, I like it but I think a small modification is needed to avoid the same issue the PR is trying to solve where Teleport would end a trainer battle.

Could `jumpifbattletype` moved inside the macro? So...
```diff
+	.macro jumpifside battler:req, side:req, equalJumpInstr:req, notEqualJumpInstr:req
+	getbattlerside \battler
+	jumpifbyte CMP_EQUAL, gBattleCommunication, \side, \equalJumpInstr
+	jumpifbyte CMP_NOT_EQUAL, gBattleCommunication, \side, \notEqualJumpInstr
+	.endm

-       getbattlerside BS_ATTACKER
-       jumpifbyte CMP_EQUAL, gBattleCommunication, B_SIDE_PLAYER, BattleScript_EffectBatonPass
+       jumpifbattletype BATTLE_TYPE_TRAINER, BattleScript_EffectBatonPass
+       jumpifside BS_ATTACKER, B_SIDE_PLAYER, BattleScript_EffectBatonPass, BattleScript_EffectTeleportTryToRunAway
-       jumpifbattletype BATTLE_TYPE_TRAINER, BattleScript_EffectBatonPass
-       jumpifbyte CMP_EQUAL, gBattleCommunication, B_SIDE_OPPONENT, BattleScript_EffectTeleportTryToRunAway
```

**Comment by AlexOn1ine** [comment]
Added macro as suggested by Lunos.

File: asm/macros/battle_script.inc [hunk 1]
```diff
@@ -2177,6 +2171,12 @@
 	jumpifbyte CMP_COMMON_BITS, gMoveResultFlags, MOVE_RESULT_NO_EFFECT, \jumpInstr
 	.endm
 
+	.macro jumpifside battler:req, side:req, equalJumpInstr:req, notEqualJumpInstr:req
+	getbattlerside \battler
+	jumpifbyte CMP_EQUAL, gBattleCommunication, \side, \equalJumpInstr
+	jumpifbyte CMP_NOT_EQUAL, gBattleCommunication, \side, \notEqualJumpInstr
+	.endm
```
Comment (AsparagusEduardo) [inline asm/macros/battle_script.inc:2178]: Uuh, aren't the jumpif macros meant to let the script continue to the next instruction if the condition doesn't pass?

**Review by AsparagusEduardo — CHANGES_REQUESTED** [review]

File: asm/macros/battle_script.inc (hunk 1 above)
Comment (LOuroboros) [reply asm/macros/battle_script.inc:2178]: They're meant to be used however the situation requires it. It's no different than doing a `call_if_eq` followed by a `call_if_ne` inside an overworld script.

In this case, the macro `jumpifside` first calls `getbattlerside` to write the side that the specified battler is on to `gBattleCommunication`, and then proceeds to jump to one battle script or another based on its value *(which is either `B_SIDE_PLAYER` or `B_SIDE_OPPONENT`)*.

Comment (AsparagusEduardo) [reply asm/macros/battle_script.inc:2178]: Yes, but the `goto BattleScript_ButItFailed` is never reached, so the second condition might as well be converted to:
```diff
-	jumpifside BS_ATTACKER, B_SIDE_PLAYER, BattleScript_EffectBatonPass, BattleScript_EffectTeleportTryToRunAway
-	goto BattleScript_ButItFailed
+	jumpifside BS_ATTACKER, B_SIDE_PLAYER, BattleScript_EffectBatonPass
+	goto BattleScript_EffectTeleportTryToRunAway
```

Comment (AlexOn1ine) [reply asm/macros/battle_script.inc:2177]: Applied

Comment (AsparagusEduardo) [reply asm/macros/battle_script.inc:2177]: Wait a second, I just realized that `BattleScript_EffectTeleportTryToRunAway` is right below it, so you can just remove that goto ^^;
image

**Review by AsparagusEduardo — APPROVED** [review]

**Pull request merged**